BParkingNano/plugins/BToKstarLLBuilder.cc

```diff
@@ -0,0 +1,207 @@
+    for(size_t ll_idx = 0; ll_idx < dileptons->size(); ++ll_idx) {
+      edm::Ptr<pat::CompositeCandidate> ll_ptr(dileptons, ll_idx);
+      edm::Ptr<reco::Candidate> l1_ptr = ll_ptr->userCand("l1");
+      edm::Ptr<reco::Candidate> l2_ptr = ll_ptr->userCand("l2");
+      int l1_idx = ll_ptr->userInt("l1_idx");
+      int l2_idx = ll_ptr->userInt("l2_idx");
+
+      pat::CompositeCandidate cand;
+      cand.setP4(ll_ptr->p4() + kstar_ptr->p4());
+      cand.setCharge( 0 ); //B0 has 0 charge
+      // save daughters - unfitted
+      cand.addUserCand("l1", l1_ptr);
+      cand.addUserCand("l2", l2_ptr);
+      cand.addUserCand("trk1", trk1_ptr);
+      cand.addUserCand("trk2", trk2_ptr);
+      cand.addUserCand("kstar", kstar_ptr);
+      cand.addUserCand("dilepton", ll_ptr);
+      // save indices
+      cand.addUserInt("l1_idx", l1_idx);
+      cand.addUserInt("l2_idx", l2_idx);
+      cand.addUserInt("trk1_idx", trk1_idx);
+      cand.addUserInt("trk2_idx", trk2_idx);
+      cand.addUserInt("kstar_idx" ,kstar_idx);
+
+      auto dr_info = min_max_dr({l1_ptr, l2_ptr, trk1_ptr, trk2_ptr});
+      cand.addUserFloat("min_dr", dr_info.first);
+      cand.addUserFloat("max_dr", dr_info.second);
+
+      // assign correct masses on tracks
+      float TRK1_MASS=K_MASS,  
+            TRK2_MASS=PI_MASS;
+      bool trk1isK=true;;
+      if (kstar_ptr->userInt("trk1isK_trk2isPi")==0 && 
+          kstar_ptr->userInt("trk1isPi_trk2isK")==1 ){
+         TRK1_MASS=PI_MASS; TRK2_MASS=K_MASS;
+         trk1isK=false;
+      }
+      // check if pass pre vertex cut
+      if( !pre_vtx_selection_(cand) ) {
+        // in case of 2 valid mass hypothesis if the first fails B cuts -> check if the other pass - in case of 2 masses check the K*-bar mass
+        if (kstar_ptr->userInt("valid_mhypothesis") <2) continue;
+        auto pp4 = kstar_ptr->polarP4();
+        pp4.SetM(kstar_ptr->userFloat("bar_mass"));
+        cand.setP4(ll_ptr->polarP4() + pp4);
+        if( !pre_vtx_selection_(cand) ) continue; //if 2nd mass fails -> skip
+        TRK1_MASS=PI_MASS; TRK2_MASS=K_MASS; //change the mass to the 2nd hypothesis
+        trk1isK=false;
+      }      
+      
+      KinVtxFitter fitter(
+        {kstars_ttracks->at(trk1_idx), kstars_ttracks->at(trk2_idx), 
+         leptons_ttracks->at(l1_idx), leptons_ttracks->at(l2_idx)},
+        {TRK1_MASS, TRK2_MASS, l1_ptr->mass(), l2_ptr->mass()},
+        { K_SIGMA, PI_SIGMA, LEP_SIGMA, LEP_SIGMA} 
+        );
+      if(!fitter.success()) continue; 
+
+      cand.setVertex( 
+        reco::Candidate::Point( 
+          fitter.fitted_vtx().x(),
+          fitter.fitted_vtx().y(),
+          fitter.fitted_vtx().z()
+          )  
+        );
+      // vertex vars
+      cand.addUserFloat("sv_chi2", fitter.chi2());
+      cand.addUserFloat("sv_ndof", fitter.dof());
+      cand.addUserFloat("sv_prob", fitter.prob());
+      cand.addUserFloat("b0fitted_kstar_mass" ,(fitter.daughter_p4(0) + fitter.daughter_p4(1)).mass());
```

Comment:
if we follow the nomenclature adopted in other modules, this userFloat could be changed to
mkstar_fullfit
and similarly for the following 3 variables

Reply:
Hi @sarafiorendi ,
Again just for claryfication: b0fitted_kstar_mass -> kstar_fullfit_mass right?
Thanks

Reply:
ok mkstar_fullfit actually sounds better... let's go with that. Thanks